Libraries/Components/MaskedView/MaskedViewIOS.ios.js

```diff
@@ -0,0 +1,169 @@
+ *       </MaskedView>
+ *     );
+ *   }
+ * }
+ * ```
+ *
+ * The above example will render a view with a blue background that fills its
+ * parent, and then mask that view with text that says "Basic Mask".
+ *
+ * The alpha channel of the view rendered by the `renderMask` prop determines how
+ * much of the view’s content and background shows through. Fully or partially
+ * opaque pixels allow the underlying content to show through but fully
+ * transparent pixels block that content.
+ *
+ */
+class MaskedViewIOS extends React.Component {
+  props: {
+    style?: StyleObj,
+    /**
+     * `MaskedView` only accepts a single child as a view to be masked.
+     */
+    children: ReactElement<*>,
+    /**
+     * Should return a React element to be rendered and applied as the
+     * mask for the child element.
+     */
+    renderMask: () => React.Element<*>,
+  };
+
+  state: {
+    maskViewNodeRef: ?mixed,
+  } = {
+    maskViewNodeRef: null,
+  };
+
+  static propTypes = {
+    ...ViewPropTypes,
+    renderMask: ReactPropTypes.func.isRequired,
+  };
+
+  _hasWarnedInvalidChildren = false;
+  _hasWarnedInvalidRenderMask = false;
+  _maskViewNodeHandle: ?number = null;
+
+  componentWillUpdate(nextProps, nextState) {
+    if (nextState.maskViewNodeRef !== this.state.maskViewNodeRef) {
+      this._maskViewNodeHandle = findNodeHandle(nextState.maskViewNodeRef);
+    }
+  }
+
+  render() {
+    if (!this.props.children || React.Children.count(this.props.children) > 1) {
+      if (!this._hasWarnedInvalidChildren) {
+        console.warn(
+          'MaskedView: MaskedView requires a single child React Element.'
+        );
+        this._hasWarnedInvalidChildren = true;
+      }
+      return null;
+    }
+
+    const target = React.Children.only(this.props.children);
+
+    if (typeof this.props.renderMask !== 'function') {
+      if (!this._hasWarnedInvalidRenderMask) {
+        console.warn(
+          'MaskedView: Invalid `renderMask` prop was passed to MaskedView. ' +
+            'Expected a function. No mask will render.'
+        );
+        this._hasWarnedInvalidRenderMask = true;
+      }
+      return <View style={this.props.style}>{target}</View>;
+    }
+
+    const maskElement = this.props.renderMask();
+    let maskElementWithRef = null;
+    if (maskElement) {
+      maskElementWithRef = cloneReferencedElement(maskElement, {
+        ref: this._handleMaskViewRef,
+      });
+    }
+
+    return (
+      <RCTMaskedView
+        style={this.props.style}
+        maskRef={this._maskViewNodeHandle}>
+        <View
+          pointerEvents="none"
+          style={[StyleSheet.absoluteFill, { opacity: 0 }]}>
+          {maskElementWithRef}
+        </View>
+        {target}
+      </RCTMaskedView>
+    );
+  }
+
+  _handleMaskViewRef = (node: React.Element<*>) => {
+    if (!node) {
+      return;
+    }
+    if (!this.state.maskViewNodeRef || this.state.maskViewNodeRef !== node) {
```

Comment:
if putting the check in the callback doesn't work then i think we should get rid of the callback and just use setState(object) so that we're either fully transactional or not at all